build/sync/plan/actions.go

```diff
@@ -0,0 +1,214 @@
+	}
+	defer dstF.Close()
+	_, err = io.Copy(dstF, srcF)
+	if err != nil {
+		return fmt.Errorf("failed to copy file %q: %v", path, err)
+	}
+	return nil
+}
+
+// OverwriteDirectoryAction is used to completely overwrite directories.
+// If the target directory exists, it will be removed first.
+type OverwriteDirectoryAction struct {
+	ActionConditions
+	Params struct {
+		// Create determines whether the target directory
+		// will be created if it does not exist.
+		Create bool `json:"create"`
+	}
+}
+
+// Run implements plan.Action.Run.
+func (a OverwriteDirectoryAction) Run(path string, setup Setup) error {
+	setup.Logf("overwriting directory %q", path)
+	src := setup.PathInRepo(TemplateRepo, path)
+	dst := setup.PathInRepo(PluginRepo, path)
+
+	dstInfo, err := os.Stat(dst)
+	switch {
+	case os.IsNotExist(err):
+		if !a.Params.Create {
+			return fmt.Errorf("path %q does not exist, not creating", dst)
+		}
+	case err != nil:
+		return fmt.Errorf("failed to check path %q: %v", dst, err)
+	default:
+		if !dstInfo.IsDir() {
+			return fmt.Errorf("path %q is not a directory", dst)
+		}
+		err = os.RemoveAll(dst)
+		if err != nil {
+			return fmt.Errorf("failed to remove directory %q: %v", dst, err)
+		}
+	}
+
+	srcInfo, err := os.Stat(src)
+	if os.IsNotExist(err) {
+		return fmt.Errorf("directory %q does not exist", src)
+	} else if err != nil {
+		return fmt.Errorf("failed to check path %q: %v", src, err)
+	}
+	if !srcInfo.IsDir() {
+		return fmt.Errorf("path %q is not a directory", src)
+	}
+
+	err = CopyDirectory(src, dst)
+	if err != nil {
+		return fmt.Errorf("failed to copy path %q: %v", path, err)
+	}
+	return nil
+}
+
+// CopyDirectory copies the directory src to dst so that after
+// a successful operation the contents of src and dst are equal.
+func CopyDirectory(src, dst string) error {
+	copier := dirCopier{dst: dst, src: src}
+	return filepath.Walk(src, copier.Copy)
+}
+
+type dirCopier struct {
+	dst string
+	src string
+}
+
+// Convert a path in the source directory to a path in the destination
+// directory.
+func (d dirCopier) srcToDst(path string) (string, error) {
+	suff := strings.TrimPrefix(path, d.src)
+	if suff == path {
+		return "", fmt.Errorf("path %q is not in %q", path, d.src)
+	}
+	return filepath.Join(d.dst, suff), nil
+}
+
+// Copy is an implementation of filepatch.WalkFunc that copies the
+// source directory to target with all subdirectories.
+func (d dirCopier) Copy(srcPath string, info os.FileInfo, err error) error {
+	if err != nil {
+		return fmt.Errorf("failed to copy directory: %v", err)
+	}
+	trgPath, err := d.srcToDst(srcPath)
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		err = os.MkdirAll(trgPath, info.Mode())
+		if err != nil {
+			return fmt.Errorf("failed to create directory %q: %v", trgPath, err)
+		}
+		err = os.Chtimes(trgPath, info.ModTime(), info.ModTime())
+		if err != nil {
+			return fmt.Errorf("failed to create directory %q: %v", trgPath, err)
+		}
```

Comment:
This will do a best effort to set the created and modified timestamps of the created directory to those of the original one. Not strictly necessary, but I thought it might be a nice-to-have.